### ISSUE TITLE
Suboptimal SELECT statements on droplets and packages

### ISSUE THREAD

**Issue by krumts** [cloudfoundry/cloud_controller_ng#783]
## Issue

The Postgres RDS instance we are using under CC is constantly under heavy load (45-60% over night, 80-95% during the day). Analyzing the SQL Statements point to a few SELECT queries from the CC.

## Context

We are running a CF system on rel 249. We are using a Postgres RDS on a db.m3.xlarge instance. The load on the DB has significantly increased over the last weeks, so I started analyzing what could be the cause of it.

I turned on statement statistics in the DB, and 3 days after I reset them, the top two queries I see are these:

```
ccdb=> select d.datname, s.calls, s.total_time, s.rows, s.query from pg_stat_statements s, pg_database d where d.oid=s.dbid order by total_time desc;

 datname  |  calls   |    total_time     |   rows    | query
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 ccdb     |  6179956 |  328972127.359034 |   5766488 | SELECT "droplets".* FROM "droplets" INNER JOIN "apps" ON ("apps"."guid" = "droplets"."app_guid") WHERE ("apps"."guid" = ?) ORDER BY "created_at" DESC, "id" DESC LIMIT ?
 ccdb     |  6179630 |  167833440.488001 |   5906915 | SELECT "packages".* FROM "packages" INNER JOIN "apps" ON ("apps"."guid" = "packages"."app_guid") WHERE ("apps"."guid" = '002dc164-ccfa-4365-b6e4-dc6a767d6038') ORDER BY "created_at" DESC, "id" DESC LIMIT 1
```

The accumulated time on the select from droplets is about 91 hours, for 3 day running.

If I ask the DB explain how these are executed, I can see that both do a table scan on the table - 1) on droplets, and 2) on packages

```
ccdb=> explain analyze SELECT "droplets".* FROM "droplets" INNER JOIN "apps" ON ("apps"."guid" = "droplets"."app_guid") WHERE ("apps"."guid" = '9ebcfd62-318d-4dde-b878-0dd940e352b7') ORDER BY "created_at" DESC, "id" DESC LIMIT 1;
                                                                   QUERY PLAN                                                                    
-------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=6573.95..6573.95 rows=1 width=1046) (actual time=86.189..86.191 rows=1 loops=1)
   ->  Sort  (cost=6573.95..6573.96 rows=5 width=1046) (actual time=86.185..86.185 rows=1 loops=1)
         Sort Key: droplets.created_at, droplets.id
         Sort Method: top-N heapsort  Memory: 25kB
         ->  Nested Loop  (cost=0.28..6573.93 rows=5 width=1046) (actual time=3.077..86.121 rows=16 loops=1)
               ->  Index Only Scan using apps_v3_guid_index on apps  (cost=0.28..8.30 rows=1 width=37) (actual time=0.024..0.028 rows=1 loops=1)
                     Index Cond: (guid = '9ebcfd62-318d-4dde-b878-0dd940e352b7'::text)
                     Heap Fetches: 1
               ->  Seq Scan on droplets  (cost=0.00..6565.57 rows=5 width=1046) (actual time=3.047..86.036 rows=16 loops=1)
                     Filter: (app_guid = '9ebcfd62-318d-4dde-b878-0dd940e352b7'::text)
                     Rows Removed by Filter: 46433
 Planning time: 0.251 ms
 Execution time: 86.277 ms
(13 rows)


 Limit  (cost=1558.80..1558.80 rows=1 width=256) (actual time=29.224..29.225 rows=1 loops=1)
   ->  Sort  (cost=1558.80..1558.81 rows=5 width=256) (actual time=29.222..29.222 rows=1 loops=1)
         Sort Key: packages.created_at, packages.id
         Sort Method: top-N heapsort  Memory: 25kB
         ->  Nested Loop  (cost=0.28..1558.78 rows=5 width=256) (actual time=0.581..29.177 rows=12 loops=1)
               ->  Index Only Scan using apps_v3_guid_index on apps  (cost=0.28..8.30 rows=1 width=37) (actual time=0.021..0.025 rows=1 loops=1)
                     Index Cond: (guid = '002dc164-ccfa-4365-b6e4-dc6a767d6038'::text)
                     Heap Fetches: 0
               ->  Seq Scan on packages  (cost=0.00..1550.43 rows=5 width=256) (actual time=0.553..29.109 rows=12 loops=1)
                     Filter: (app_guid = '002dc164-ccfa-4365-b6e4-dc6a767d6038'::text)
                     Rows Removed by Filter: 45081
 Planning time: 0.194 ms
 Execution time: 29.273 ms
(13 rows)
```
Running these repeatedly I can see execution time going up to 200ms for a single select, depending on the load.
Running a select on droplets using some of the indices (e.g. on created_at) I get 0.000xxx ms

Looking from the other side - the access log of the nginx on CC, one can correlate that certain API calls which result in many (over 100) of these queries take long to complete - 10 to 15 sec.
```
api.cf.sap.hana.ondemand.com - [20/Feb/2017:12:15:13 +0000] "GET /v2/spaces?inline-relations-depth=1 HTTP/1.1" 200 549591 "-" "Apache-HttpClient/4.5.2 (Java/1.8.0_102)" 10.0.0.53, 10.0.65.92, 10.0.65.8 vcap_request_id:16adb19f-c221-43ce-6469-b3bb61cc7c5f::f69fa484-2d76-4f9b-9ddd-d50d3053c41a response_time:10.281
```
Back to the queries - I had a look at the droplets table - currently we have about 45.000 entries in both droplets and packages, which is not much for a DB table I thing. Most of them however have a state "EXPIRED":

```
ccdb=> select count(state), state from droplets group by state;
 count |  state  
-------+---------
 29837 | EXPIRED
  2222 | FAILED
 14414 | STAGED
(3 rows)

ccdb=> select count(state), state from packages group by state;
 count |       state       
-------+-------------------
 28732 | EXPIRED
 16322 | READY
    61 | FAILED
     1 | PROCESSING_UPLOAD
(4 rows)
```
What does "EXPIRED" mean? Many of the entries date June 2016. I doubt that our users still use these.
So combining both - it looks like the load we observe is mostly caused by queries which 1) do a table scan on 2) a table which seems to grow over time, without some cleanup possibility?

## Expected result

Excellent performance with low CPU usage ;-)

## Current result

See above - we have pretty high load on the DB, slow responses on some of the API calls.

## Possible Fix

I think adding an index on the app_guid may help, but I didn't dare to add it manually, as I don't know what schema changes future CC versions will bring, and  I am afraid I may cause an issue during a future update of the system. I may consider adding one, if you ask me to.

I think it would be nice to have some way to cleanup the tables from obsolete entries. Can one actually identify entries as obsolete? I've seen some pull requests related to cleanup of the blobstore, but I think the DB tables are not touched.

I don't really understand why the two tables are joined. From what I see in the first statement only values from the droplets table are selected, and the where clause is also on a droplets column. Is it to make sure that there is also an entry in the apps table? I guess I need to understand SQL better.

Help is appreciated!

**Comment by krumts** [comment]
Just some additional info I forgot to add:
the third top query we see is related I think to https://github.com/cloudfoundry/hm9000/issues/22 - a-few-screens-long chain of nested selects:
```
 ccdb     |   135587 |  143988569.252002 |   5423480 | SELECT * FROM (SELECT "droplets".*, "apps"."guid" AS "x_foreign_key_x" FROM "droplets" INNER JOIN "apps" ON ("apps"."guid" = "droplets"."app_guid") WHERE ('922e191b-4620-42e5-9653-b5f37d2daa86' = "apps"."guid") ORDER BY "created_at" DESC, "id" DESC LIMIT 1) AS "t1" UNION ALL SELECT * FROM (SELECT "droplets".*, "apps"."guid" AS "x_foreign_key_x" FROM "droplets" INNER JOIN "apps" ON ("apps"."guid" = "droplets"."app_guid") WHERE ('b4e78915-3a53-4a3c-9685-3a71ccc14f85' = "apps"."guid") ORDER BY "created_at" DESC, "id" DESC LIMIT 1) AS "t1" UNION ALL SELECT * FROM (SELECT "droplets".*, "apps"."guid" AS "x_foreign_key_x" FROM "droplets" INNER JOIN "apps" ON ("apps"."guid" = "droplets"."app_guid") WHERE ('be4f2b49-f753-4644-b559-b6843b2d78ba' = "apps"."guid") ORDER BY "created_at" DESC, "id" DESC LIMIT 1) AS "t1" UNION ALL SELECT * FROM (SELECT "droplets".*, "apps"."guid" AS "x_foreign_key_x" FROM "droplets" INNER JOIN "apps" ON ("apps"."guid" = "droplets"."
. . .
```

**Comment by jberkhahn** [comment]
Hello @krumts 

We are aware of the poor performance of some of these queries - GET /v2/spaces?inline-relations-depth=1 in particular as it basically queries the state of the entire system. This was one of the motivating factors behind relationships in v3, but that's still a ways off yet. We'll take a look at adding an apps_guid index and see what kind of performance increase we can get.

As an aside, it looks like you're still running on DEAs. Support for them has been deprecated and they'll likely be removed in the near future, so you may wish to consider switching to Diego.

Thanks,
@jberkhahn
CF CAPI Team

**Comment by krumts** [comment]
Hi @jberkhahn
Thanks for looking into the issue I opened.
Meanwhile I did the following experiment - cloned the DB (so that I have the same data) and added an index on droplets(app_guid) in my test DB.
Testing with a few guids shows quite good improvements for the most frequently called SELECT statements we observed:
- without index:
```
ccdb=> explain analyze SELECT "droplets".* FROM "droplets" INNER JOIN "apps" ON ("apps"."guid" = "droplets"."app_guid") WHERE ("apps"."guid" = '7b807839-de83-48bb-ada3-4a42f2852aa3') ORDER BY "created_at" DESC, "id" DESC LIMIT 1;
 Limit  (cost=7155.44..7155.44 rows=1 width=1028) (actual time=39.025..39.026 rows=1 loops=1)
   ->  Sort  (cost=7155.44..7156.26 rows=326 width=1028) (actual time=39.023..39.023 rows=1 loops=1)
         Sort Key: droplets.created_at, droplets.id
         Sort Method: top-N heapsort  Memory: 25kB
         ->  Nested Loop  (cost=0.28..7153.81 rows=326 width=1028) (actual time=0.099..38.473 rows=336 loops=1)
               ->  Index Only Scan using apps_v3_guid_index on apps  (cost=0.28..8.30 rows=1 width=37) (actual time=0.029..0.032 rows=1 loops=1)
                     Index Cond: (guid = '7b807839-de83-48bb-ada3-4a42f2852aa3'::text)
                     Heap Fetches: 1
               ->  Seq Scan on droplets  (cost=0.00..7142.25 rows=326 width=1028) (actual time=0.063..37.735 rows=336 loops=1)
                     Filter: (app_guid = '7b807839-de83-48bb-ada3-4a42f2852aa3'::text)
                     Rows Removed by Filter: 50874
 Planning time: 0.197 ms
 Execution time: 39.078 ms
```
- with index:
```
ccdb=> explain analyze SELECT "droplets".* FROM "droplets" INNER JOIN "apps" ON ("apps"."guid" = "droplets"."app_guid") WHERE ("apps"."guid" = '7b807839-de83-48bb-ada3-4a42f2852aa3') ORDER BY "created_at" DESC, "id" DESC LIMIT 1;
                                                                     QUERY PLAN                                                                     
----------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=1096.23..1096.23 rows=1 width=1028) (actual time=2.102..2.103 rows=1 loops=1)
   ->  Sort  (cost=1096.23..1097.05 rows=326 width=1028) (actual time=2.099..2.099 rows=1 loops=1)
         Sort Key: droplets.created_at, droplets.id
         Sort Method: top-N heapsort  Memory: 25kB
         ->  Nested Loop  (cost=15.22..1094.60 rows=326 width=1028) (actual time=0.164..1.612 rows=336 loops=1)
               ->  Index Only Scan using apps_v3_guid_index on apps  (cost=0.28..8.30 rows=1 width=37) (actual time=0.030..0.032 rows=1 loops=1)
                     Index Cond: (guid = '7b807839-de83-48bb-ada3-4a42f2852aa3'::text)
                     Heap Fetches: 1
               ->  Bitmap Heap Scan on droplets  (cost=14.94..1083.04 rows=326 width=1028) (actual time=0.127..0.902 rows=336 loops=1)
                     Recheck Cond: (app_guid = '7b807839-de83-48bb-ada3-4a42f2852aa3'::text)
                     Heap Blocks: exact=330
                     ->  Bitmap Index Scan on droplets_by_app_guid  (cost=0.00..14.86 rows=326 width=0) (actual time=0.085..0.085 rows=336 loops=1)
                           Index Cond: (app_guid = '7b807839-de83-48bb-ada3-4a42f2852aa3'::text)
 Planning time: 0.219 ms
 Execution time: 2.155 ms
```
Therefore I think adding the index as you proposed would help.

About DEA usage - we are aware of the fact that it is deprecated. We already use Diego as default runtime, but still offer the possibility to use DEA. Thanks for the hint anyway.

Last remark - what about the other question I asked - we have currently about 50.000 entries in the droplets and packages tables, and about 7000 apps. Most of the entries in droplets and packages have state EXPIRED. Should these tables (droplets, packages) be cleaned up? Is there a way to do so?

Thanks again for the help!

**Comment by zrob** [comment]
@krumts We believe this commit should help https://github.com/cloudfoundry/cloud_controller_ng/commit/485dcc6a4789af71952e42304eb09f2b017f09d5

it will be available in cf-release 254

if you think you'll get a chance to test it and update us that'd be awesome

**Comment by krumts** [comment]
@zrob thanks for providing a fix!
I can't think of an easy way to get a clone of out productive system (i.e. same number of droplets, same load) to test what the end effect of the change is. For the testing if an index helps (i.e. my previous comment), I simply restored the DB to a new instance and tested there (just with a Postgres cli). Your changes however, include also some caching in the CC, and to test I need a copy of the whole system where  I then only update the CC.

I can definitely give feedback once we are on release 254.  We are trying to regularly update to the latest one, so I guess it would be about two weeks after it is released.
Is this fair enough?

**Comment by zrob** [comment]
sounds good!

**Comment by aashah** [comment]
With respect to having the table cleaned up, there is a story prioritized (currently at the top) for cleaning up expired droplet records: https://www.pivotaltracker.com/story/show/140529637.

**Comment by krumts** [comment]
Hey @aashah this sounds great! Thanks a lot!
I hope the improvements will be of benefit for many long-lived CF installations.

**Comment by zrob** [comment]
Closing for now.  Please feel free to re-open if you do not see any improvements after moving to cf 254.

**Comment by krumts** [comment]
Thanks for the fixes! We just updated to rel. 256 and I can confirm the expected improvements.
The CPU load on the DB dropped significantly, and the performance of some of the end user activities (e.g. listing spaces) improved significantly.
I checked that both of the SELECTs I listed initially run with an index search instead of with a table scan.

I still have one question related to the cleanup job. If I could follow the git history properly, v 256 (which we have) should already contain the changes. Looking in the DB I can confirm this for the droplets - I see currently only 406 entries with state EXPIRED (instead of ~30.000 previously). However, for the packages I still get ~30.000 packages with state EXPIRED.
Is there some where a description under which circumstances the entries should be cleaned up from the DB? Is it worth to open another issue to discuss this concrete observation (in case it seems strange to you)?
